### PR TITLE
[Sparse] Sparse matrix subtraction

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Support sparse matrix subtraction: A - B

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/sparse/elementwise_op.py [hunk 1]
```diff
@@ -64,32 +66,38 @@ def sub(A: Union[DiagMatrix], B: Union[DiagMatrix]) -> Union[DiagMatrix]:
...
-    >>> B = diag(torch.arange(10, 13))
-    >>> sub(A, B)
-    DiagMatrix(val=tensor([-9, -9, -9]),
-               shape=(3, 3))
+    >>> row = torch.tensor([1, 0, 2])
+    >>> col = torch.tensor([0, 1, 2])
+    >>> val = torch.tensor([10, 20, 30])
+    >>> A = from_coo(row, col, val)
+    >>> B = diag(torch.arange(1, 4))
+    >>> add(A, B)
```
Comment (mufeili) [inline python/dgl/sparse/elementwise_op.py:96]: sub, please also try the example locally

Comment (mufeili) [inline python/dgl/sparse/elementwise_op.py:96]:
```suggestion
    >>> sub(A, B)
```

File: tests/pytorch/sparse/test_elementwise_op.py [hunk 2]
```diff
@@ -2,18 +2,22 @@
 import sys
 
 import backend as F
+
+import dgl.sparse
```
Comment (mufeili) [inline tests/pytorch/sparse/test_elementwise_op.py:6]: `import dgl.sparse as sp`?

File: tests/pytorch/sparse/test_elementwise_op.py [hunk 3]
```diff
@@ -2,18 +2,22 @@
...
 import pytest
 import torch
-
-from dgl.sparse import add, diag, from_coo, from_csc, from_csr
+from dgl.sparse import add, diag, from_coo, from_csc, from_csr, sub
```
Comment (mufeili) [inline tests/pytorch/sparse/test_elementwise_op.py:9]: Unify L6 and L9?

File: python/dgl/sparse/elementwise_op_diag.py [hunk 4]
```diff
@@ -48,15 +48,17 @@ def diag_add(
...
     ----------
     D1 : DiagMatrix
         Diagonal matrix
-    D2 : DiagMatrix
-        Diagonal matrix
+    D2 : DiagMatrix or SparseMatrix
+        Diagonal matrix or sparse matrix
 
     Returns
     -------
```
Comment (frozenbugs) [inline python/dgl/sparse/elementwise_op_diag.py:64]: Update return type.
```
    DiagMatrix or SparseMatrix
        Diagonal matrix or sparse matrix, same as D2
```

File: python/dgl/sparse/elementwise_op_diag.py [hunk 5]
```diff
@@ -77,6 +79,56 @@ def diag_sub(D1: DiagMatrix, D2: DiagMatrix) -> DiagMatrix:
...
+
+    Examples
+    --------
+    >>> D1 = diag(torch.arange(1, 4))
+    >>> D2 = diag(torch.arange(10, 13))
+    >>> D2 - D1
+    DiagMatrix(val=tensor([-9, -9, -9]),
+    shape=(3, 3))
+    """
+    if isinstance(D2, DiagMatrix):
```
Comment (frozenbugs) [inline python/dgl/sparse/elementwise_op_diag.py:119]: What about just return -(D1-D2)?

File: tests/pytorch/sparse/test_elementwise_op.py (hunk 2 above)
Comment (frozenbugs) [reply tests/pytorch/sparse/test_elementwise_op.py:6]: `import dgl.sparse as dglsp`

Let's follow the conversion.

```
./tests/pytorch/sparse/test_reduction.py:import dgl.sparse as dglsp
./examples/sparse/sign.py:import dgl.sparse as dglsp
./examples/sparse/sgc.py:import dgl.sparse as dglsp
./examples/sparse/gcnii.py:import dgl.sparse as dglsp
./examples/sparse/c_and_s.py:import dgl.sparse as dglsp
./examples/sparse/hypergraphatt.py:import dgl.sparse as dglsp
./examples/sparse/appnp.py:import dgl.sparse as dglsp
./examples/sparse/gcn.py:import dgl.sparse as dglsp
./examples/sparse/graph_transformer.py:import dgl.sparse as dglsp
./examples/sparse/gat.py:import dgl.sparse as dglsp
./examples/sparse/hgnn.py:import dgl.sparse as dglsp
./examples/sparse/twirls.py:import dgl.sparse as dglsp
```

File: python/dgl/sparse/elementwise_op_diag.py (hunk 4 above)
Comment (BarclayII) [reply python/dgl/sparse/elementwise_op_diag.py:64]: Done.

File: python/dgl/sparse/elementwise_op_diag.py [hunk 6]
```diff
@@ -49,20 +49,22 @@ def diag_add(
...
-    D2 : DiagMatrix
-        Diagonal matrix
+    D2 : DiagMatrix or SparseMatrix
+        Diagonal matrix or sparse matrix
 
     Returns
     -------
     DiagMatrix
-        Diagonal matrix
+        Diagonal matrix, same as D2
```
Comment (frozenbugs) [inline python/dgl/sparse/elementwise_op_diag.py:67]: the result of D - A should be sparse, right?

Comment (BarclayII) [reply python/dgl/sparse/elementwise_op_diag.py]: fixed

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**